pennylane_forest/qpu.py

```diff
@@ -67,11 +93,154 @@ def __init__(self, device, *, shots=1024, active_reset=False, load_qc=True, **kw
         aspen_match = re.match(r"Aspen-\d+-([\d]+)Q", device)
         num_wires = int(aspen_match.groups()[0])
 
-        super(QVMDevice, self).__init__(
-            num_wires, shots, **kwargs
-        )  # pylint: disable=bad-super-call
+        super().__init__(num_wires, shots, **kwargs)
 
         if load_qc:
             self.qc = get_qc(device, as_qvm=False, connection=self.connection)
+        else:
+            self.qc = get_qc(device, as_qvm=True, connection=self.connection)
```

Comment:
Yes, to both questions. I basically gave the ability to initialize a QVM here so that it's easy to test if readout correction works for this device, and we wouldn't have to carry out QPU experiments every time we make a change.

Reply:
Nice! I think that makes a lot of sense in terms of device initialization.

Reply:
To be clear, you can use pyquil's operator_estimation module to run something like `9q-qvm`, but if you try to do something like `dev = qml.device('forest.qpu', device='9q-qvm', load_qc=False)` you'll get an error.

We could also potentially go the route of also giving the QVMDevice the ability to possess some specified readout error model, and the ability to correct it.

Reply:
Yep. This might also be a good time to evaluate the `QVMDevice` vs `QPUDevice` abstraction. Does it make sense considering how the QVM and QPUs are treated in PyQuil? Or would it make more sense to merge it into a single device?

Reply:
I gave this some thought. I think it might ultimately make sense to move both devices into a single `QCDevice` (QC = QuantumComputer), or some such thing. At the same time, I have a strong suspicion that if we try to do this now, we'll run into a bunch of problems or even breaking changes. My preference would be to merge the two devices gently, over probably a few PRs.

Once this PR gets merged, I'll make an issue about qubit indexing that I ran into here (and is alluded to in the example notebook, where we ask the user to select a lattice that has the qubit `0`). I think resolving that would be a good starting point as a guide on how to merge the two devices into one.

Reply:
> I think it might ultimately make sense to move both devices into a single QCDevice (QC = QuantumComputer), or some such thing.

I agree. Perhaps we should also rename the `device=` keyword argument to `qc=`, or even `get_qc=` to better match pyquil.

```python
dev = qml.device("forest.qc", qc="Aspen-4-4Q-A")
```

Reply:
Yeah, that would make sense. But again, I think we should do that in a future PR once we're fairly confident merging the two won't cause any problems. I can think of at least one immediate issue, related to the fact qubits on our physical QPUs are not necessarily indexed 0, ..., num_qubits - 1.

Reply:
Actually, here's the issue in a nutshell: `num_wires` is something that's specified by the QPU lattice that you instantiate, and pennylane doesn't seem to allow applications on qubits with indices beyond `num_wires`. For example, `Aspen-4-3Q-A` has 3 qubits which are indexed `10`, `11` and `17`. However, if I try to do a simple rotation on qubit 17 as in the following code
```
import pyquil
import pennylane as qml
from pennylane import numpy as np

dev_qpu = qml.device('forest.qpu', device='Aspen-4-3Q-A', load_qc=False)
qubit = 17

@qml.qnode(dev_qpu)
def circuit():
    qml.RY(np.pi/2, wires=qubit)
    return qml.expval(qml.PauliX(qubit))

circuit()
```
I get the error msg (from `pennylane/qnode.py`): `QuantumFunctionError: Operation RY applied to invalid wire 17 on device with 3 wires.`.

Reply:
Yep. I had a play around with this, and my thinking was, as long as the mapping was available (i.e., `mode_map = dict(zip(range(num_wires)), [10, 11, 17])`, we can do this mapping from within `apply()`.

When we designed PennyLane, we made the assumption that QPUs would always use consecutive numbering :laughing:

Reply:
Lol in all honesty, I actually don't know why we number our qubits that way!

You can certainly access the qubits for a `QuantumComputer` by doing `qc.qubits()`, e.g.
```
from pyquil import get_qc
qc = get_qc('Aspen-4-3Q-A', as_qvm=True)
qc.qubits()
```
prints `[10, 11, 17]`.

Let me try tinkering with `apply()` in the morning (or maybe a bit later), and get back to you about this then. Maybe this won't be an issue after all.

Reply:
I think two independent solutions came out of this comment thread :laughing: Happy to move them to new issues/PRs to allow this to be merged in.